utils/inputs/pagination.go

```diff
@@ -0,0 +1,162 @@
+}
+
+// GetOrderBy - create the order by expression and parameters for pagination
+func (p Pagination) GetOrderBy() string {
+	var (
+		statement string
+	)
+
+	for _, po := range p.Order {
+		if statement != "" {
+			// not the first statement
+			statement += ", "
+		}
+		statement += fmt.Sprintf("%s ", po.Attribute)
+
+		if po.Direction != "" {
+			if po.Direction == Ascending {
+				statement += " ASC "
+			} else if po.Direction == Descending {
+				statement += " DESC "
+			}
+		}
+	}
+	return statement
+}
+
+// Validate - implementation of validatable interface
+func (p *Pagination) Validate(ctx context.Context) error {
+	var errs = new(errorutils.MultiError)
+	if p.Page < 0 {
+		errs.Append(errors.New("page value must be greater than or equal to 0"))
+	}
+	if p.Items <= 0 {
+		errs.Append(errors.New("items value must be greater than 0"))
+	}
+
+	// get allowed values for order from context, if nothing allow all values
+	if okOrder, ok := ctx.Value(appctx.PaginationOrderOptionsCTXKey).(map[string]bool); ok {
+		for _, o := range p.Order {
+			if _, ok := okOrder[o.Attribute]; !ok {
+				errs.Append(fmt.Errorf("order parameter '%s' is not allowed", o.Attribute))
+			}
+		}
+	}
+
+	if errs.Count() > 0 {
+		return errs
+	}
+
+	return nil
+}
+
+// Decode - implementation of  decodable interface
+func (p *Pagination) Decode(ctx context.Context, v []byte) error {
+	u, err := url.Parse(string(v))
+	if err != nil {
+		return fmt.Errorf("failed to parse pagination parameters: %w", err)
+	}
+
+	// get the query string parameters
+	q := u.Query()
+
+	if q.Get("page") == "" {
+		p.Page = 0
+	} else {
+		p.Page, err = strconv.Atoi(q.Get("page"))
+		if err != nil {
+			return fmt.Errorf("failed to parse pagination page parameter: %w", err)
+		}
+	}
+
+	if q.Get("items") == "" {
+		p.Items = 10
+	} else {
+		p.Items, err = strconv.Atoi(q.Get("items"))
+		if err != nil {
+			return fmt.Errorf("failed to parse pagination items parameter: %w", err)
+		}
+	}
+
+	for _, v := range q["order"] {
+		parts := strings.Split(v, ".")
+		po := PageOrder{}
+		if parts[0] != "" {
+			po.Attribute = parts[0]
+		}
+		if len(parts) > 1 && parts[1] != "" {
+			if string(Ascending) == strings.ToUpper(parts[1]) {
+				po.Direction = Ascending
+			}
+			if string(Descending) == strings.ToUpper(parts[1]) {
+				po.Direction = Descending
+			}
+		}
+		p.Order = append(p.Order, po)
+	}
+	p.RawOrder = q["order"]
+
+	return nil
+}
+
+// NewPagination - create a new Pagination struct and populate from url and order options
+func NewPagination(ctx context.Context, url string, orderOptions ...string) (*Pagination, error) {
+	var (
+		pagination = new(Pagination)
+		order      = map[string]bool{}
+	)
+	// allowed pagination ordering attributes
+	for _, o := range orderOptions {
+		order[o] = true
+	}
+	ctx = context.WithValue(ctx, appctx.PaginationOrderOptionsCTXKey, order)
+
+	if err := DecodeAndValidate(ctx, pagination, []byte(url)); err != nil {
+		fmt.Println("!!!!! err: ", err)
```

Comment:
Should we log this error rather than printing?

Reply:
in next commit.